src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/WebWorker.cs

```diff
@@ -0,0 +1,124 @@
+                        try
+                        {
+                            if (t.IsFaulted)
+                                tcs.SetException(t.Exception);
+                            else if (t.IsCanceled)
+                                tcs.SetCanceled();
+                            else
+                                tcs.SetResult(t.Result);
+
+                            JSHostImplementation.UninstallWebWorkerInterop();
+
+                        }
+                        catch (Exception e)
+                        {
+                            Environment.FailFast("WebWorker.RunAsync failed", e);
+                        }
+                    }, TaskScheduler.Current);
+                }
+                catch (Exception e)
+                {
+                    Environment.FailFast("WebWorker.RunAsync failed", e);
+                }
+
+            });
+            JSHostImplementation.SetHasExternalEventLoop(t);
+            t.Start();
+            return tcs.Task;
+        }
+
+        public static Task RunAsyncVoid(Func<Task> body)
+        {
+            var tcs = new TaskCompletionSource();
+            var t = new Thread(() =>
+            {
+                try
+                {
+                    JSHostImplementation.InstallWebWorkerInterop(true);
+                    Task res = body();
+                    // This code is exiting thread main() before all promises are resolved.
+                    // the continuation is executed by setTimeout() callback of the thread.
+                    res.ContinueWith((t) =>
+                    {
+                        try
+                        {
+                            if (t.IsFaulted)
+                                tcs.SetException(t.Exception);
+                            else if (t.IsCanceled)
+                                tcs.SetCanceled();
+                            else
+                                tcs.SetResult();
+
+                            JSHostImplementation.UninstallWebWorkerInterop();
+                        }
+                        catch (Exception e)
+                        {
+                            Environment.FailFast("WebWorker.RunAsync failed", e);
+                        }
+                    }, TaskScheduler.Current);
+                }
+                catch (Exception e)
+                {
+                    Environment.FailFast("WebWorker.RunAsync failed", e);
+                }
+            });
+            JSHostImplementation.SetHasExternalEventLoop(t);
+            t.Start();
+            return tcs.Task;
+        }
+
+        public static Task Run(Action body)
+        {
+            var tcs = new TaskCompletionSource();
+            var t = new Thread(() =>
+            {
+                try
+                {
+                    JSHostImplementation.InstallWebWorkerInterop(false);
+                    body();
+                    tcs.SetResult(); // TODO how to pass this to the original thread ?
```

Comment:
I coverted this back to draft and changed implementation strategy. 
- I will install the interop support from managed code of WebWorker.
- I will track interop and async interop separately.

This place in code currently doesn't work.  The result is not propagated to the outer thread and the code hangs on the outer await. I guess I need to `Send`  some wakeup there.